modules/apps/router/module.go

```diff
@@ -0,0 +1,300 @@
+}
+
+// OnChanOpenConfirm implements the IBCModule interface
+func (am AppModule) OnChanOpenConfirm(ctx sdk.Context, portID, channelID string) error {
+	// call underlying app's OnChanOpenConfirm callback.
+	return am.app.OnChanOpenConfirm(ctx, portID, channelID)
+}
+
+// OnChanCloseInit implements the IBCModule interface
+func (am AppModule) OnChanCloseInit(ctx sdk.Context, portID, channelID string) error {
+	// TODO: Unescrow all remaining funds for unprocessed packets
+	return am.app.OnChanCloseInit(ctx, portID, channelID)
+}
+
+// OnChanCloseConfirm implements the IBCModule interface
+func (am AppModule) OnChanCloseConfirm(ctx sdk.Context, portID, channelID string) error {
+	// TODO: Unescrow all remaining funds for unprocessed packets
+	return am.app.OnChanCloseConfirm(ctx, portID, channelID)
+}
+
+// OnRecvPacket implements the IBCModule interface.
+func (am AppModule) OnRecvPacket(ctx sdk.Context, packet channeltypes.Packet, relayer sdk.AccAddress) ibcexported.Acknowledgement {
+	ack := ibcexported.Acknowledgement(channeltypes.NewResultAcknowledgement([]byte{byte(1)}))
+
+	var data transfertypes.FungibleTokenPacketData
+	if err := transfertypes.ModuleCdc.UnmarshalJSON(packet.GetData(), &data); err != nil {
+		ack = channeltypes.NewErrorAcknowledgement("cannot unmarshal ICS-20 transfer packet data")
+	}
+
+	// only attempt the application logic if the packet data
+	// was successfully decoded
+	if ack.Success() {
+		receiver, finalDest, port, channel, err := ParseIncomingTransferField(data.Receiver)
+		if finalDest == "" && port == "" && channel == "" && err == nil {
+			ack = am.app.OnRecvPacket(ctx, packet, relayer)
+		} else if err != nil {
+			ack = channeltypes.NewErrorAcknowledgement("cannot unmarshal ICS-20 transfer packet data")
+		} else {
+			if err := am.keeper.OnRecvPacket(ctx, packet, data, receiver, finalDest, port, channel); err != nil {
+				ack = channeltypes.NewErrorAcknowledgement(err.Error())
+			}
+		}
+
+	}
+
+	ctx.EventManager().EmitEvent(
+		sdk.NewEvent(
+			transfertypes.EventTypePacket,
+			sdk.NewAttribute(sdk.AttributeKeyModule, types.ModuleName),
+			sdk.NewAttribute(transfertypes.AttributeKeyReceiver, data.Receiver),
+			sdk.NewAttribute(transfertypes.AttributeKeyDenom, data.Denom),
+			sdk.NewAttribute(transfertypes.AttributeKeyAmount, fmt.Sprintf("%d", data.Amount)),
+			sdk.NewAttribute(transfertypes.AttributeKeyAckSuccess, fmt.Sprintf("%t", ack.Success())),
+		),
+	)
+
+	return nil
+}
+
+// OnAcknowledgementPacket implements the IBCModule interface
+func (am AppModule) OnAcknowledgementPacket(ctx sdk.Context, packet channeltypes.Packet, acknowledgement []byte, relayer sdk.AccAddress) error {
+	return am.app.OnAcknowledgementPacket(ctx, packet, acknowledgement, relayer)
+}
+
+// OnTimeoutPacket implements the IBCModule interface
+func (am AppModule) OnTimeoutPacket(ctx sdk.Context, packet channeltypes.Packet, relayer sdk.AccAddress) error {
+	return am.app.OnTimeoutPacket(ctx, packet, relayer)
```

Comment:
ditto we must send new packet here to refund original sender in routing case

Reply:
This feature would leave the user on the intermediate chain at their refund address in the case where packet times out or fails.